### PR TITLE
Intersection schema generation is order-dependent

### PR DESCRIPTION
**Pull request by altearius**
Addresses #602 and #597

### CODE REVIEW

File: src/applySchemaTyping.ts
Comment (bcherny) [inline src/applySchemaTyping.ts:9]: `$intersection` and `$types` may conflict with schema properties (in JSON-Schema, users can define custom properties). Instead, can you use Symbols (similar to `Parent`)? This name collision could also be good to add as a test case.

https://github.com/bcherny/json-schema-to-typescript/blob/31993def993b610ba238d3024260129e31ddc371/src/types/JSONSchema.ts#L43

File: src/applySchemaTyping.ts
Comment (bcherny) [inline src/applySchemaTyping.ts:18]: What about `title`?

File: src/parser.ts
Comment (bcherny) [inline src/parser.ts:60]:
```suggestion
    throw new ReferenceError('Expected intersection schema')
```

File: src/types/JSONSchema.ts
Comment (bcherny) [inline src/types/JSONSchema.ts:76]: Nit: please type this explicitly to keep it simple.
```suggestion
  $types: readonly [SchemaType, ...SchemaType[]]
```

**Review by bcherny — CHANGES_REQUESTED** [review]
Thank you for the contribution!

Your approach looks good to me. I tried a more surgical fix in https://github.com/bcherny/json-schema-to-typescript/pull/605/files, but like that your approach is more generic and cleaner.

A few minor things before we merge this:

1. Would you mind splitting your PR into two commits: one to add the new test cases, and one to make code changes and change the output of your test cases? That would make it more clear how your changes change the output of the test cases.
2. Would you mind simplifying the naming in your cases (eg. `A` instead of `Level2A`)? (example: https://github.com/bcherny/json-schema-to-typescript/pull/605/commits/1252285c9f934aff7e1875965e779866bab228c4)

File: src/applySchemaTyping.ts
Comment (altearius) [reply src/applySchemaTyping.ts]: This suggestion significantly improved the code quality, thank you.

File: src/applySchemaTyping.ts
Comment (altearius) [reply src/applySchemaTyping.ts:27]: I missed `title` as it was not handled by the original [`maybeStripNameHints` function][1], though I have now added it.

[1]: https://github.com/bcherny/json-schema-to-typescript/blob/31993def993b610ba238d3024260129e31ddc371/src/utils.ts#L341-L352 'maybeStripNameHints'

**Comment by altearius** [comment]
Thank you for the code review! This is ready for a second pass.

File: test/__snapshots__/test/test.ts.md
Comment (bcherny) [inline test/__snapshots__/test/test.ts.md:1876]: I think I missed this earlier -- this is a case where the output in #605 is better than #603. I wonder if there's a way to improve this output, so we don't generate `A1` and `B1`, and instead just use `Base` directly.

Like I did here: https://github.com/bcherny/json-schema-to-typescript/pull/605/files#diff-34615f003eb49f90c3044b8fafebe3fe6b162e290b9b724082a6d72f58cfa4afR1876

File: test/__snapshots__/test/test.ts.md
Comment (bcherny) [inline test/__snapshots__/test/test.ts.md:1928]: And here. This output is cleaner, if you can achieve something similar with your approach: https://github.com/bcherny/json-schema-to-typescript/pull/605/files#diff-34615f003eb49f90c3044b8fafebe3fe6b162e290b9b724082a6d72f58cfa4afR1927

File: test/__snapshots__/test/test.ts.md
Comment (bcherny) [inline test/__snapshots__/test/test.ts.md:448971]: Nice

**Review by bcherny — CHANGES_REQUESTED** [review]
Looking better!

The output is a nice improvement across a number of tests. I wonder if there might be a way to improve the output more for the test cases you added.

To help debug, try adding `export const only = true` to your test case, then running `VERBOSE=1 npm test` to enable debug tracing.

File: test/__snapshots__/test/test.ts.md
Comment (altearius) [reply test/__snapshots__/test/test.ts.md:1876]: Thanks! I took a try at that in 7d3f495 by pulling schema's pre-existing `allOf` rule into the intersection. Would you mind taking a third pass and let me know if I missed anything?

File: test/__snapshots__/test/test.ts.md
Comment (bcherny) [inline test/__snapshots__/test/test.ts.md:447878]: This seems to have regressed

File: test/__snapshots__/test/test.ts.md
Comment (bcherny) [inline test/__snapshots__/test/test.ts.md:225127]: This looks like a regression. Asked about it here to understand the expected behavior: https://github.com/json-schema-org/json-schema-spec/issues/1529

File: test/__snapshots__/test/test.ts.md
Comment (altearius) [reply test/__snapshots__/test/test.ts.md:447878]: Just a heads up: I haven't lost interest in moving this forward. I've had some deadlines come up, but I plan to return to this over the weekend.

Comment (altearius) [reply test/__snapshots__/test/test.ts.md:447878]: Hello! Thank you for your patience. I've taken a look at this regression and I'm not sure what the expected behavior should be. Could you advise?

This is what I observe, simplified to the relevant bits.

Given this schema:

```yaml
title: JSON schema for NPM package.json files
type: object
patternProperties:
  '^_':
    description: Any property starting with _ is valid.
properties:
  # etc, etc, etc, actual properties clipped
```

The original code produced this structure:

```ts
export type JSONSchemaForNPMPackageJsonFiles =
  JSONSchemaForNPMPackageJsonFiles1 & JSONSchemaForNPMPackageJsonFiles2;

export type JSONSchemaForNPMPackageJsonFiles1 = { [k: string]: unknown; };

export interface JSONSchemaForNPMPackageJsonFiles2 {␊
  /**
   * Any property starting with _ is valid.
   * 
   * This interface was referenced by \`JSONSchemaForNPMPackageJsonFiles2\`'s
   * JSON-Schema definition via the \`patternProperty\` "^_".
   */
  [k: string]: any;
}
```

This branch produces a different structure:

```ts
export type JSONSchemaForNPMPackageJsonFiles =
  { [k: string]: unknown; } &
  {
    /**
     * Any property starting with _ is valid.
     * 
     * This interface was referenced by \`undefined\`'s JSON-Schema definition
     * via the \`patternProperty\` "^_".
     */
    [k: string]: any;
  };
```

While I agree that the change in the comment from `JSONSchemaForNPMPackageJsonFiles2` to `undefined` is unfortunate, this behavior is in line with other instances in which this comment is generated due to the presence of `patternProperties` on the root schema.

For instance, here is the current output of the `intersection.2.ts` test on the `master` branch, which has a similar scenario:

```ts
export type Intersection = (A & B) & {
  c: string;
  d: string;
  /**
   * This interface was referenced by \`undefined\`'s JSON-Schema definition
   * via the \`patternProperty\` "^x-".
   */
  [k: string]: unknown;
};

export interface A {
  a?: string;
  [k: string]: unknown;
}

export interface B {
  b?: string;
  [k: string]: unknown;
}
```

I would argue that neither behavior is ideal, and an improvement might be to suppress the comment entirely if the interface is not named.

I can make this change, but I worry that it may not be what you have in mind, and that it feels like something that should be a separate PR.

Could you please provide direction on how this might be resolved?

File: test/__snapshots__/test/test.ts.md
Comment (altearius) [reply test/__snapshots__/test/test.ts.md:225127]: Given this schema:

```yaml
type: object
required: [ruleType, timeWindowSize]

oneOf:
  - $ref: '#/definitions/ActiveConnectionsNotInAllowedRange'
  # ...etc...
  - $ref: '#/definitions/UnauthorizedOperationsNotInAllowedRange'

properties:
  ruleType:
    type: string
    enum: [ TimeWindowCustomAlertRule ]

  timeWindowSize:
    type: string
```

The old behavior is:

```ts
export interface DeviceSecurityGroupProperties {
  // Note: does **not** include the ruleType or timeWindowSize properties
  timeWindowRules?:
    | (
        | ActiveConnectionsNotInAllowedRange
        // ...etc...
        | UnauthorizedOperationsNotInAllowedRange
      )[]
    | string;
  [k: string]: unknown;
}

export type TimeWindowCustomAlertRule = {
  ruleType: "TimeWindowCustomAlertRule";
  timeWindowSize: string;
  [k: string]: unknown;
} & (
  | ActiveConnectionsNotInAllowedRange
  // ...etc...
  | UnauthorizedOperationsNotInAllowedRange
);
```

The new behavior is:

```ts
export interface DeviceSecurityGroupProperties {
  // Note: Now includes ruleType and timeWindowSize.
  timeWindowRules?: TimeWindowCustomAlertRule[] | string;
}

export type TimeWindowCustomAlertRule = {
  ruleType: "TimeWindowCustomAlertRule";
  timeWindowSize: string;
  [k: string]: unknown;
} & (
  | ActiveConnectionsNotInAllowedRange
  // ...etc...
  | UnauthorizedOperationsNotInAllowedRange
);
```

I would argue that the new behavior is more nearly correct. My reasoning is based on the definition of the `oneOf` keyword in the [JSON Schema spec][1]:

> 10.2.1.3. oneOf
>
> This keyword's value MUST be a non-empty array. Each item of the array MUST be a valid JSON Schema.
>
> An instance validates successfully against this keyword if it validates successfully against exactly one schema defined by this keyword's value.

This definition does not address the possibility that `oneOf` may be combined with additional keywords, such as `properties`. However, one reasonable interpretation is that the `properties` keyword should be applied in addition to the `oneOf` keyword. In such a case, both `oneOf` and `properties` (or other keywords) would all apply.

This interpretation aligns with [section 10.1][3]:

> 10.1 Keyword Independence
>
> Schema keywords typically operate independently, without affecting each other's outcomes.

This suggests that `oneOf` and `properties` should _both_ affect the output.

This interpretation is also in line with the [Factoring Schemas][2] example, which demonstrates how one schema:

```yaml
oneOf:
  - { type: number, multipleOf: 5 }
  - { type: number, multipleOf: 3 }
```

Can be factored into a schema that combines `oneOf` with additional keywords (`type`, in this case, but I presume an example with `properties` would be equally valid):

```yaml
type: number
oneOf:
  - { multipleOf: 5 }
  - { multipleOf: 3 }
```

I realize that other interpretations of the `oneOf` schema are possible, but this is what feels most intuitive to me.  Please let me know if other changes need to be made.

[1]: https://json-schema.org/draft/2020-12/json-schema-core#section-10.2.1.3 'oneOf'
[2]: https://json-schema.org/understanding-json-schema/reference/combining#factoringschemas 'Factoring Schemas'
[3]: https://json-schema.org/draft/2020-12/json-schema-core#name-keyword-independence 'Keyword Independence'

File: test/__snapshots__/test/test.ts.md
Comment (bcherny) [reply test/__snapshots__/test/test.ts.md]: That seems reasonable -- we can improve this as a follow-up.

File: test/__snapshots__/test/test.ts.md
Comment (bcherny) [reply test/__snapshots__/test/test.ts.md]: I think that's a reasonable interpretation, but as you said, there may be other interpretations. Implicit intersection schemas like this make me scratch my head sometimes. I *think* your interpretation is correct, and the new behavior is correct, but since it's a subtle breaking change that will affect the largest consumers of JSTT, I want to double check with JSON-Schema maintainers first.

I bumped the Github issue to see is someone is willing to clarify.

**Comment by bcherny** [comment]
Published as v15.0.0.

**Pull request merged**
